### PR TITLE
fix(deps): k8s dependencies to v0.31.0

### PR DESCRIPTION
**Pull request by tnthornton**
### Description of your changes
General dependency upgrade to try to get better memory efficiencies with JSON marshal/unmarshal.

I have:

- [X] Read and followed Upbound's [contribution process](https://git.io/fj2m9).
- [X] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested
Ran make test locally.

### CODE REVIEW

File: internal/graph/model/apiextensions.go [hunk 1]
```diff
@@ -171,7 +162,6 @@ func GetComposition(cmp *extv1.Composition) Composition {
 			},
 			WriteConnectionSecretsToNamespace: cmp.Spec.WriteConnectionSecretsToNamespace,
 		},
-		Status: GetCompositionStatus(cmp.Status),
```
Comment (avalanche123) [inline internal/graph/model/apiextensions.go:174]: does this change the schema?

File: internal/unstructured/composite.go [hunk 2]
```diff
@@ -264,3 +265,18 @@ func (c *Composite) SetEnvironmentConfigReferences(refs []corev1.ObjectReference
...
 	_ = fieldpath.Pave(c.Object).SetValue("spec.environmentConfigRefs", filtered)
 }
+
+// SetObservedGeneration of this composite resource claim.
+func (c *Composite) SetObservedGeneration(generation int64) {
+	status := &xpv1.ObservedStatus{}
+	_ = fieldpath.Pave(c.Object).GetValueInto("status", status)
+	status.SetObservedGeneration(generation)
+	_ = fieldpath.Pave(c.Object).SetValue("status.observedGeneration", status.ObservedGeneration)
+}
```
Comment (avalanche123) [inline internal/unstructured/composite.go:275]: not sure where this is used? is this related to the change in a schema object?

**Review by avalanche123 — APPROVED** [review]
a few questions

File: internal/unstructured/composite.go (hunk 2 above)
Comment (tnthornton) [reply internal/unstructured/composite.go:275]: This is because the Composition API changed and this function is expected now.

File: internal/graph/model/apiextensions.go (hunk 1 above)
Comment (tnthornton) [reply internal/graph/model/apiextensions.go:174]: Not sure. I did deploy these changes to prod and didn't see an issue 🤷

File: cmd/xgql/main.go [hunk 3]
```diff
@@ -191,7 +191,11 @@ func main() { //nolint:gocyclo
 	// until they are resolved (e.g. the user is granted the RBAC access they
 	// need) or the cache expires.
 	//nolint:reassign
-	utilruntime.ErrorHandlers = []func(error){func(err error) { log.Debug("Kubernetes runtime error", "err", err) }}
+	utilruntime.ErrorHandlers = []utilruntime.ErrorHandler{
+		func(ctx context.Context, err error, msg string, keysAndValues ...interface{}) {
+			log.Debug("Kubernetes runtime error", "err", err)
```
Comment (pohly) [inline cmd/xgql/main.go:196]: The error is not a complete description of the problem and may be nil. See https://github.com/kubernetes/kubernetes/issues/135349 and https://github.com/kubernetes/kubernetes/pull/135419 for some helper code that you might want to use.
